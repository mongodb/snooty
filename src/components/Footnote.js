import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import ComponentFactory from './ComponentFactory';
import { getNestedValue } from '../utils/get-nested-value';
import { intersperse } from '../utils/intersperse';
import FootnoteContext from './footnote-context';

const Footnote = ({ nodeData: { children, id, name }, ...rest }) => {
  const { footnotes } = useContext(FootnoteContext);
  const ref = name || id.replace('id', '');
  const label = getNestedValue([ref, 'label'], footnotes);
  const footnoteReferences = footnotes[ref] ? footnotes[ref].references : [];
  const footnoteReferenceNodes = footnoteReferences.map((footnote, index) => (
    <a className="fn-backref" href={`#ref-${footnote}`}>
      {index + 1}
    </a>
  ));
  return (
    <table className="docutils footnote" frame="void" id={ref} rules="none">
      <colgroup>
        <col className="label" />
      </colgroup>
      <tbody valign="top">
        <tr>
          <td className="label">
            [
            {footnoteReferenceNodes.length > 1 ? (
              label
            ) : (
              <a className="fn-backref" href={`#ref-${footnoteReferences[0]}`}>
                {label}
              </a>
            )}
            ]
          </td>
          <td>
            {footnoteReferenceNodes.length > 1 && <em>({intersperse(footnoteReferenceNodes)})</em>}{' '}
            {children.map((child, index) => (
              <ComponentFactory {...rest} nodeData={child} key={index} parentNode="footnote" />
            ))}
          </td>
        </tr>
      </tbody>
    </table>
  );
};

Footnote.propTypes = {
  nodeData: PropTypes.shape({
    children: PropTypes.arrayOf(PropTypes.object),
    id: PropTypes.string.isRequired,
    name: PropTypes.string,
  }).isRequired,
};

export default Footnote;
