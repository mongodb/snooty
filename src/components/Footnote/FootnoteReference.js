import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import FootnoteContext from './footnote-context';
import { getNestedValue } from '../../utils/get-nested-value';

const FootnoteReference = ({ nodeData: { id, refname } }) => {
  const { footnotes } = useContext(FootnoteContext);

  // the nodeData originates from docutils, and may be incorrect for
  // anonymous footnoteReferences originating from included files -- docutils
  // appears to assign IDs within the included files before they are collated

  const ref = refname || id.replace('id', '');
  const uid = refname ? `${refname}-${id}` : id;
  return (
    <a className="footnote-reference" href={`#footnote-${ref}`} id={`ref-${uid}`}>
      [{getNestedValue([ref, 'label'], footnotes) || ref}]
    </a>
  );
};

FootnoteReference.propTypes = {
  nodeData: PropTypes.shape({
    id: PropTypes.string.isRequired,
    refname: PropTypes.string,
  }).isRequired,
};

export default FootnoteReference;
