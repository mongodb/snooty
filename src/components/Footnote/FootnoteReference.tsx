import React, { useContext } from 'react';
import { useDarkMode } from '@leafygreen-ui/leafygreen-provider';
import { css, cx } from '@leafygreen-ui/emotion';
import { palette } from '@leafygreen-ui/palette';
import { getNestedValue } from '../../utils/get-nested-value';
import FootnoteContext from './footnote-context';

const refStyles = css`
  color: ${palette.blue.light1};

  &:visited {
    color: ${palette.purple.base};
  }
`;

/**
 * Component used to show a clickable reference to footnote on page
 * scrolls to referenced element by id property
 */

type NodeData = {
  id: string;
  refname: string;
};

type FootnoteReferenceProps = {
  nodeData: NodeData;
};

type FootnotesContextType = {
  footnotes: Record<string, any>;
};

const FootnoteReference = ({ nodeData }: FootnoteReferenceProps) => {
  const { footnotes } = useContext(FootnoteContext) as FootnotesContextType;
  const { darkMode } = useDarkMode();

  const { id, refname } = nodeData;

  const ref = refname || id.replace('id', '');
  const uid = refname ? `${refname}-${id}` : id;

  return (
    <a
      className={cx('footnote-reference header-buffer', darkMode && refStyles)}
      href={`#footnote-${ref}`}
      id={`ref-${uid}`}
    >
      [{getNestedValue([ref, 'label'], footnotes) || ref}]
    </a>
  );
};

export default FootnoteReference;
