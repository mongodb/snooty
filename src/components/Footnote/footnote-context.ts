import { createContext } from 'react';

interface Footnote {
  references: string[];
  label: string;
}

interface FootnoteContextType {
  footnotes: Record<string, Footnote>;
}

const FootnoteContext = createContext<FootnoteContextType>({
  footnotes: {},
});

export default FootnoteContext;
