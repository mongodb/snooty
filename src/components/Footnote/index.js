import React, { useContext } from 'react';
import PropTypes from 'prop-types';
import { css } from '@emotion/react';
import FootnoteContext from './footnote-context';
import ComponentFactory from '../ComponentFactory';
import { getNestedValue } from '../../utils/get-nested-value';
import { intersperse } from '../../utils/intersperse';

const tableStyling = css`
  border: 0;
  border-collapse: collapse;
  margin: 24px 0;
  font-size: 14px;
  line-height: 24px;

  tbody tr td div.highlight pre {
    background-color: inherit;
  }

  :target {
    background-color: #ffa;
  }
`;

const tdStyling = css`
  border: 0;
  padding: 11px 5px 12px;
`;

const Footnote = ({ nodeData: { children, id, name }, ...rest }) => {
  const { footnotes } = useContext(FootnoteContext);
  const ref = name || id.replace('id', '');
  const label = getNestedValue([ref, 'label'], footnotes);
  const uid = name ? `${name}-` : '';
  const footnoteReferences = footnotes && footnotes[ref] ? footnotes[ref].references : [];
  const footnoteReferenceNodes = footnoteReferences.map((footnote, index) => (
    <a href={`#ref-${uid}${footnote}`} key={index}>
      {index + 1}
    </a>
  ));
  return (
    <table css={tableStyling} frame="void" id={`footnote-${ref}`} rules="none">
      <colgroup>
        <col />
      </colgroup>
      <tbody valign="top">
        <tr>
          <td css={tdStyling}>
            [{footnoteReferenceNodes.length !== 1 ? label : <a href={`#ref-${uid}${footnoteReferences[0]}`}>{label}</a>}
            ]
          </td>
          <td css={tdStyling}>
            {footnoteReferenceNodes.length > 1 && <em>({intersperse(footnoteReferenceNodes)})</em>}{' '}
            {children.map((child, index) => (
              <ComponentFactory {...rest} nodeData={child} key={index} parentNode="footnote" />
            ))}
          </td>
        </tr>
      </tbody>
    </table>
  );
};

Footnote.propTypes = {
  nodeData: PropTypes.shape({
    children: PropTypes.arrayOf(PropTypes.object),
    id: PropTypes.string.isRequired,
    name: PropTypes.string,
  }).isRequired,
};

export default Footnote;
