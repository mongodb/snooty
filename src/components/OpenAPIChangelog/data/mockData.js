export const mockChangelog = [
  {
    date: '2024-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'removed',
            changes: [
              {
                change: 'resource version 2023-01-01 was removed, latest available resource version is 2023-02-01',
                changeCode: 'resource-version-removed',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-01',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/alertConfigs/{alertConfigId}',
        httpMethod: 'PUT',
        operationId: 'updateAlertConfiguration',
        tag: 'Alert Configurations',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'release',
            changes: [
              {
                change: 'resource version added.',
                changeCode: 'resource-version-added',
                backwardCompatible: false,
              },
              {
                change:
                  "for the 'query' request parameter 'clusterName', default value was changed from 'default' to 'cluster0'",
                changeCode: 'request-parameter-default-value-changed',
                backwardCompatible: false,
              },
              {
                change: "added the optional property 'links' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'deprecate',
            sunsetDate: '2024-02-01',
            changes: [
              {
                change: 'deprecated by resource version 2023-02-01 and marked for removal for 2024-02-01.',
                changeCode: 'resource-version-deprecated',
                backwardCompatible: false,
              },
            ],
          },
        ],
      },
      {
        path: '/api/atlas/v2/groups/{groupId}/accessList',
        httpMethod: 'POST',
        operationId: 'createProjectIpAccessList',
        tag: 'Project IP Access List',
        versions: [
          {
            version: '2023-01-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change: "added the optional property 'name' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
              {
                change: "added the optional property 'comment' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
  {
    date: '2023-02-22',
    paths: [
      {
        path: '/api/atlas/v2/groups/byName/{groupName}',
        httpMethod: 'GET',
        operationId: 'getProjectByName',
        tag: 'Projects',
        versions: [
          {
            version: '2023-02-01',
            stabilityLevel: 'stable',
            changeType: 'update',
            changes: [
              {
                change:
                  "added the optional property 'withDefaultAlertsSettings' in the response with the '200' status.",
                changeCode: 'response-optional-property-added',
                backwardCompatible: true,
              },
            ],
          },
        ],
      },
    ],
  },
];

export const mockDiff = [
  {
    path: '/api/atlas/v2/federationSettings/{federationSettingsId}/connectedOrgConfigs',
    httpMethod: 'GET',
    operationId: 'listConnectedOrgConfigs',
    tag: 'Federated Authentication',
    changeType: 'update',
    changes: [
      {
        change: "removed the non-success response with the status '400'.",
        changeCode: 'response-non-success-status-removed',
        backwardCompatible: false,
      },
      {
        change: "added the new optional 'query' request parameter 'envelope'.",
        changeCode: 'new-optional-request-parameter',
        backwardCompatible: true,
      },
    ],
  },
  {
    path: '/api/atlas/v2/groups/{groupId}/accessList',
    httpMethod: 'POST',
    operationId: 'createProjectIpAccessList',
    tag: 'Project IP Access List',
    changeType: 'removed',
    changes: [
      {
        change: "removed the required property 'name' from the response with the '200' status.",
        changeCode: 'response-required-property-removed',
        backwardCompatible: false,
      },
    ],
  },
];

export const mockIndex = {
  runDate: '2023-04-03',
  runId: '20230403105948111',
  specRevision: '422ab47f864909549362a7b39688404c82e2540b',
  specRevisionShort: '422ab47f864',
  versions: ['2023-01-01', '2023-02-01', '2023-08-11~preview'],
};
