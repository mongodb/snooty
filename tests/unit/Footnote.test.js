import React from 'react';
import { shallow } from 'enzyme';
import Footnote from '../../src/components/Footnote';

// data for this component
import mockData from './data/Footnote.test.json';

it('renders correctly', () => {
  const tree = shallow(<Footnote nodeData={mockData} />);
  expect(tree).toMatchSnapshot();
});
