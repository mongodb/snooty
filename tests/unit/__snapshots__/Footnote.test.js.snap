// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<table
  className="docutils footnote"
  frame="void"
  id="1"
  rules="none"
>
  <colgroup>
    <col
      className="label"
    />
  </colgroup>
  <tbody
    valign="top"
  >
    <tr>
      <td
        className="label"
      >
        [
        <a
          className="fn-backref"
          href="#ref-undefined"
        />
        ]
      </td>
      <td>
         
        <ComponentFactory
          key="0"
          nodeData={
            Object {
              "children": Array [
                Object {
                  "position": Object {
                    "start": Object {
                      "line": 108,
                    },
                  },
                  "type": "text",
                  "value": "Numerical footnote.",
                },
              ],
              "position": Object {
                "start": Object {
                  "line": 108,
                },
              },
              "type": "paragraph",
            }
          }
          parentNode="footnote"
        />
      </td>
    </tr>
  </tbody>
</table>
`;
